test/e2e/tls_util_test.go

```diff
@@ -142,3 +159,73 @@ func TestOnlyAppSecretExist(t *testing.T) {
 		t.Fatalf("expect %v, but got %v", tlsutil.ErrCANotFound.Error(), err.Error())
 	}
 }
+
+// TestOnlyCAExist ensures that at the case where only the CA exists in the cluster;
+// GenerateCert can retrieve the CA and uses it to create a new application secret.
+func TestOnlyCAExist(t *testing.T) {
+	f := framework.Global
+	ctx := f.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+	namespace, err := ctx.GetNamespace()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = f.KubeClient.CoreV1().ConfigMaps(namespace).Create(caConfigMap)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = f.KubeClient.CoreV1().Secrets(namespace).Create(caSecret)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cg := tlsutil.NewSDKCertGenerator(f.KubeClient)
+	// Use Pod as a dummy runtime object for the CR input of GenerateCert().
+	mCR := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind: crKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      crName,
+			Namespace: namespace,
+		},
+	}
+	appSvc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "app-service",
+			Namespace: namespace,
+		},
+	}
+	appSecret, _, _, err := cg.GenerateCert(mCR, appSvc, ccfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// check if appSecret has the correct fields.
+	if appSecretName != appSecret.Name {
+		t.Fatalf("expect the secret name %v, but got %v", appSecretName, appSecret.Name)
+	}
+	if namespace != appSecret.Namespace {
+		t.Fatalf("expect the secret namespace %v, but got %v", namespace, appSecret.Namespace)
+	}
+	if v1.SecretTypeTLS != appSecret.Type {
+		t.Fatalf("expect the secret type %v, but got %v", v1.SecretTypeTLS, appSecret.Type)
+	}
+	if _, ok := appSecret.Data[v1.TLSCertKey]; !ok {
+		t.Fatalf("expect the secret to have the data field %v, but got none", v1.TLSCertKey)
+	}
+	if _, ok := appSecret.Data[v1.TLSPrivateKeyKey]; !ok {
+		t.Fatalf("expect the secret to have the data field %v, but got none", v1.TLSPrivateKeyKey)
+	}
+
+	// check if appSecret exists in k8s cluster.
+	appSecretFromCluster, err := f.KubeClient.CoreV1().Secrets(namespace).Get(appSecretName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	// check if appSecret returned from GenerateCert is the same as the one that exists in the k8s.
+	if !reflect.DeepEqual(appSecret, appSecretFromCluster) {
+		t.Fatalf("%+v differs %+v", appSecret, appSecretFromCluster)
```

Comment:
Change this error message to `Expected: %+v \n Got: %+v`. Otherwise it's not clear what was expected.

Reply:
sure

Reply:
I am going to use `"expect %+v, but got %+v"` to be consistent.